System Stepping implemented as Resource

dmlary opened the pull request:
# Objective

Add interactive system debugging capabilities to bevy, providing step/break/continue style capabilities to running system schedules.

* Original implementation: #8063
    - `ignore_stepping()` everywhere was too much complexity
* Schedule-config & Resource discussion: #8168
    - Decided on selective adding of Schedules & Resource-based control

## Solution
Created `Stepping` Resource.  This resource can be used to enable stepping on a per-schedule basis.  Systems within schedules can be individually configured to:
* AlwaysRun: Ignore any stepping state and run every frame
* NeverRun: Never run while stepping is enabled
    - this allows for disabling of systems while debugging
* Break: If we're running the full frame, stop before this system is run

Stepping provides two modes of execution that reflect traditional debuggers:
* Step-based: Only execute one system at a time
* Continue/Break: Run all systems, but stop before running a system marked as Break

### Demo
https://user-images.githubusercontent.com/857742/233630981-99f3bbda-9ca6-4cc4-a00f-171c4946dc47.mov

Breakout has been modified to use Stepping.  The game runs normally for a couple of seconds, then stepping is enabled and the game appears to pause.  A list of Schedules & Systems appears with a cursor at the first System in the list.  The demo then steps forward full frames using the spacebar until the ball is about to hit a brick.  Then we step system by system as the ball impacts a brick, showing the cursor moving through the individual systems.  Finally the demo switches back to frame stepping as the ball changes course.


### Limitations
Due to architectural constraints in bevy, there are some cases systems stepping will not function as a user would expect.

#### Event-driven systems
Stepping does not support systems that are driven by `Event`s as events are flushed after 1-2 frames. Although game systems are not running while stepping, ignored systems are still running every frame, so events will be flushed. 

This presents to the user as stepping the event-driven system never executes the system. It does execute, but the events have already been flushed. 

This can be resolved by changing event handling to use a buffer for events, and only dropping an event once all readers have read it. 

The work-around to allow these systems to properly execute during stepping is to have them ignore stepping: `app.add_systems(event_driven_system.ignore_stepping())`. This was done in the breakout example to ensure sound played even while stepping. 

#### Conditional Systems
When a system is stepped, it is given an opportunity to run.  If the conditions of the system say it should not run, it will not. 

Similar to Event-driven systems, if a system is conditional, and that condition is only true for a very small time window, then stepping the system may not execute the system.  This includes depending on any sort of external clock.

This exhibits to the user as the system not always running when it is stepped. 

A solution to this limitation is to ensure any conditions are consistent while stepping is enabled. For example, all systems that modify any state the condition uses should also enable stepping.

#### State-transition Systems
Stepping is configured on the per-`Schedule` level, requiring the user to have a `ScheduleLabel`. 

To support state-transition systems, bevy generates needed schedules dynamically. Currently it’s very difficult (if not impossible, I haven’t verified) for the user to get the labels for these schedules. 

Without ready access to the dynamically generated schedules, and a resolution for the `Event` lifetime, **stepping of the state-transition systems is not supported**

---

## Changelog
- `Schedule::run()` updated to consult `Stepping` Resource to determine which Systems to run each frame
- Added `Schedule.label` as a `BoxedSystemLabel`, along with supporting `Schedule::set_label()` and `Schedule::label()` methods
    - `Stepping` needed to know which `Schedule` was running, and prior to this PR, `Schedule` didn't track its own label
    - Would have preferred to add `Schedule::with_label()` and remove `Schedule::new()`, but this PR touches enough already
- Added calls to `Schedule.set_label()` to `App` and `World` as needed
- Added `Stepping` resource
- Added `Stepping::begin_frame()` system to `MainSchedulePlugin`
    - Run before `Main::run_main()`
    - Notifies any `Stepping` Resource a new render frame is starting
    
## Migration Guide
- Add a call to `Schedule::set_label()` for any custom `Schedule`
    - This is only required if the `Schedule` will be stepped

dmlary commented:
Opening this so I can use github to do my own code-review.

Please hold off on reviewing it until I pull it out of draft mode.

dmlary commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1305 @@
+        // if the stepping frame cursor is pointing at this schedule, we'll run
+        // the schedule with the current stepping action.  If this is not the
+        // cursor schedule, we'll run the schedule with the waiting action.
+        let Cursor {
+            schedule: schedule_index,
+            system: start,
+        } = self.cursor;
```
Probably don't need to destructure here.  See what it looks like without.

dmlary commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1305 @@
+            Some(i) => self.cursor.system = i,
+            None => {
+                let index = schedule_index + 1;
+                println!(
+                    "Stepping::skipped_systems(): next schedule {} -> {}",
+                    schedule_index, index,
+                );
```
remove

dmlary commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1305 @@
+    node_ids: Vec<NodeId>,
+
+    /// changes to system behavior that should be applied the next time
+    /// [`ScheduleState::progress()`] is called
```
update function name in comment.

dmlary commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1305 @@
+
+    // apply system behavior updates by looking up the node id of the system in
+    // the schedule, and updating `systems`
+    fn apply_updates(&mut self, schedule: &Schedule) {
```
`apply_behavior_updates()`, and rename `updates` to `behavior_updates`.

dmlary commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1305 @@
+    // apply system behavior updates by looking up the node id of the system in
+    // the schedule, and updating `systems`
+    fn apply_updates(&mut self, schedule: &Schedule) {
+        for (node_id, system) in schedule.systems().unwrap() {
```
Add comment explaining that systems may be present multiple times in a schedule, so we iterate through all systems and lookup any changes for that system's behavior and apply them to the node.

dmlary commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1305 @@
+        }
+        self.updates.clear();
+
+        println!("apply_updates(): {:?}", self.behaviors);
```
remove

dmlary commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1305 @@
+    behaviors: HashMap<NodeId, SystemBehavior>,
+
+    /// order of NodeIds in the schedule
+    node_ids: Vec<NodeId>,
```
Expand comment:
This is a cached copy of `SystemExecutable.system_ids`.  We need it available here to be accessed by `Stepping::cursor()` so we can return `NodeId`s to the caller.

dmlary commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1305 @@
+
+        // if our NodeId list hasn't been populated, copy it over from the
+        // schedule
+        if self.node_ids.len() != schedule.systems_len() {
+            self.node_ids = schedule.executable().system_ids.clone();
+        }
```
### Reaching into SystemExecutable for NodeIds
This feels ugly to me, but I don't have a better fast approach for getting this list.  I'm avoiding iterating through `Schedule::systems()` to extract NodeIds because that's much slower.

Suggestions are welcomed here.

### Improved Schedule update detection
Right now this "works", but it relies on the fact that systems can only be added to schedules.

There's an alternate approach here where we update `Schedule` to have a `generation` or `update_at` field, and we detect changes from that.  This could resolve the other comment about `systems_len()`.

This would look like:
```rust
if schedule.generation > self.last_schedule_update {
    self.node_ids = schedule.executable().system_ids.clone();
    self.last_schedule_update = schedule.generation;
}
```

I won't make this change without a reviewer requesting it.

dmlary commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1305 @@
+                .behaviors
+                .get(&node_id)
+                .unwrap_or(&SystemBehavior::Continue);
+            println!(
+                "skipped_systems(): systems[{}], pos {}, Action::{:?}, Behavior::{:?}, {}",
+                i,
+                pos,
+                action,
+                behavior,
+                system.name()
+            );
```
See if you can get this message to only show up during testing, otherwise remove debug message

dmlary commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1305 @@
+                // is marked as NeverRun, add it to the skip list
+                (_, SystemBehavior::NeverRun) => {
+                    skip.insert(i);
+                    // always advance the position past this sytem
```
move this comment up to the match arm, and clean it up some:
Also advance our cursor past this system if it is our current position

dmlary commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1305 @@
+                // if we're waiting, no other systems besides AlwaysRun should
+                // be run, so add systems to the skip list
+                (Action::Waiting, _) => skip.insert(i),
+                // If we're stepping, the remaining behaviors don't matter,
+                // we're only going to run the system at our cursor.  Any system
+                // not at our cursor is skipped.  Once we encounter the system
+                // at the cursor, we'll move the cursor up, and switch to
+                // Waiting, to skip remaining systems.
```
Any system prior to our cursor will be skipped.

dmlary commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1305 @@
+                (Action::RunAll, _) => unreachable!(),
+            }
+
+            if i == pos && action != Action::Waiting {
+                pos += 1;
+            }
```
explain this update to the cursor position

dmlary commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1305 @@
+        );
+    }
+
+    // Helper to verify cursor position
+    /*
+    macro_rules! assert_cursor_at {
+        ($stepping:expr, $schedule:expr, $system:expr) => {
+            let type_id = IntoSystem::into_system($system).type_id();
+            let node_id = $schedule
+                .systems()
+                .unwrap()
+                .find(|t| t.1.type_id() == type_id)
+                .unwrap()
+                .0;
+            let label = $schedule.label().as_ref().unwrap();
+            assert_eq!($stepping.cursor(), Some((label.dyn_clone(), node_id)));
+        };
+    } */
```
remove

dmlary commented on examples/games/stepping.rs:
```diff
@@ -0,0 +1,277 @@
+
+impl SteppingPlugin {
+    /// Initialize the plugin to step the schedules specified in `labels`
+    pub fn for_schedules(labels: Vec<Box<dyn ScheduleLabel>>) -> SteppingPlugin {
```
`BoxedScheduleLabel`

dmlary commented on examples/games/stepping.rs:
```diff
@@ -0,0 +1,277 @@
+/// executing schedule is removed from the [`Schedules`] resource while it is
+/// being run.
+#[derive(Debug, Hash, PartialEq, Eq, Clone, ScheduleLabel)]
+struct Debug;
```
`DebugSchedule`?

dmlary commented on examples/games/stepping.rs:
```diff
@@ -0,0 +1,277 @@
+        for (node_id, system) in systems {
+            if system.name().starts_with("bevy") {
+                always_run.push((label.dyn_clone(), node_id));
+                println!("skipping {:?}", system.name());
```
remove debug message

dmlary commented on examples/games/stepping.rs:
```diff
@@ -0,0 +1,277 @@
+                },
+            ));
+
+            println!("found system: {:?}/{}\n", label, system.name());
```
change to `debug!` message

dmlary commented on examples/games/stepping.rs:
```diff
@@ -0,0 +1,277 @@
+    let (ui, mut text, vis) = ui.single_mut();
+    match (vis, stepping.is_enabled()) {
+        (Visibility::Hidden, true) => {
+            println!("showing UI");
```
remove these debug messages.

dmlary commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1305 @@
+                    }
+                },
+                Update::SetBehavior(label, system, behavior) => {
+                    match self.schedule_states.get_mut(&label) {
+                        Some(state) => state.set_behavior(system, behavior),
+                        None => {
+                            warn!(
+                                "stepping is not enabled for schedule {:?}; use
+                    `.add_stepping({:?})` to enable stepping",
+                                label, label
+                            );
+                        }
+                    }
+                }
```
meh, gotta put the macro outside of the `impl` block.  I'm leaving it as is for now.

dmlary commented on examples/games/stepping.rs:
```diff
@@ -0,0 +1,277 @@
+                println!("skipping {:?}", system.name());
+                continue;
+            }
+
```
Automatically skipping event-reader systems requires changing `Stepping::build_ui()` to use `mut World`.  That change really complicates the method.  For now I'm going to hold off on adding this capability to the demos.

dmlary commented:
> I'm not a fan of the extra branch and added parameter to Schedule::run. It's complexity that anyone that wants to make a new executor type has to handle.   Is it possible to just make this a new executor type?

### Custom Stepping Executor

There are some challenges to this approach.  First of all, not all systems in a schedule may be stepping, which means we need to support both multi-threaded & single threaded executors to ensure intra-system behavior is present while stepping.  I didn't want to duplicate the existing executors, and have to keep them in sync with the real executors.

Also, schedule executor is configured on a per-schedule basis.  This is possible, but keep in mind, the user doesn't actually have the `Schedule` structs; they have to be dug out of the `Schedules` resource, and only if they're not running in the schedule itself.

I honestly think this is a more complex approach for bevy users, and want to avoid it.

### Complexity of `skip_systems` for custom executors

The short answer here is that a custom executor doesn't have to support stepping if the bevy ones do.  The complexity of implementing it depends on the executor itself.

### Performance

The performance aspect is something I've looked into.  For many_foxes, I'm not seeing a FPS hit; I performed the same performance testing I describe in the **Performance Details** for #8063.  (side-note: the bevy main branch has dropped something like 25-30 fps since the last time I ran these tests).

That said, although it's not observable, yes, there absolutely is a performance impact here.  It's not just a branch, there's the whole `Stepping` resource lookup, and branch into that even if stepping is disabled.  I'm open to adding a `#[cfg(...)]` option prefixing that block, so it's an explicitly-enabled feature.  I would recommend we use something other than just `#[cfg(debug)]` as we all end up debugging in the release build at some point.

I'm not really sure how to add a custom feature, and if this is the right use here.

alice-i-cecile commented:
Yeah, I think that adding this to a feature flag is the way: this is complexity and a performance cost that I don't want to ship in a final product. Definitely on by default though: make things nice for beginners.

This is a [good introduction](https://doc.rust-lang.org/cargo/reference/features.html#features): you'll need one in `bevy_ecs`, `bevy_app` and probably [`bevy_internal`](https://github.com/bevyengine/bevy/blob/main/crates/bevy_internal/Cargo.toml).

dmlary commented:
> Yeah, I think that adding this to a feature flag is the way: this is complexity and a performance cost that I don't want to ship in a final product. Definitely on by default though: make things nice for beginners.
> 
> This is a [good introduction](https://doc.rust-lang.org/cargo/reference/features.html#features): you'll need one in `bevy_ecs`, `bevy_app` and probably [`bevy_internal`](https://github.com/bevyengine/bevy/blob/main/crates/bevy_internal/Cargo.toml).

Thanks for the references.  I got it sorted and figured out the workspace feature particularities.  Added feature `bevy_debug_stepping`.

The `Stepping` resource remains even when the feature is disabled, but all the hooks are removed.  If the user tries `Stepping::enable()` they'll get an error message.

dmlary commented on crates/bevy_ecs/src/schedule/schedule.rs:
```diff
@@ -266,6 +288,11 @@ impl Schedule {
     }
 
+    /// Returns the [`SystemSchedule`].
+    pub fn executable(&self) -> &SystemSchedule {
```
Ok, `Stepping` does need access to the `SystemSchedule`.  Switched this to `pub(crate)`.

dmlary commented:
@cart i know you’ve got a lot on your plate, but could you do a quick review of this PR when you get a moment?

I’d like to get it into the 0.11 release, but I will need time to make any changes you require before the release.

nerdachse commented on crates/bevy_ecs/src/schedule/schedule.rs:
```diff
@@ -300,6 +327,36 @@ impl Schedule {
+        Ok(iter)
+    }
+
+    /// Returns the number of systems in this schedule.
+    pub fn systems_len(&self) -> usize {
+        if !self.executor_initialized {
+            self.graph.systems.len()
+        } else {
+            self.executable.systems.len()
+        }
+    }
```
I know it's probably unpopular as there are reasons not to use too long names, but I tend to prefer longer function names instead of a clarifying comment, so I would go with:

`pub fn systems_in_schedule`

dmlary replied:
I am not opposed to long names at all.  My only reluctance with the name `systems_in_schedule()` is from the name alone I'd expect to get a vector of systems back, which `systems()` already does.  I ended up on `systems_len()` because I was doing `systems().len()`, but wanted to reduce the performance overhead.

Other options: `systems_count`, `number_of_systems`, `total_systems`, `count_systems`, `schedule_size` ... oh I kind of like `schedule_size`, but `size` feels like bytes to me.

All that said, I'm happy to rename to `systems_in_schedule()`.

rodolphito replied:
`system_len` is fine, but if it cant be that then im partial to `systems_count`.

dmlary replied:
I'll leave this as system_len().

dmlary replied:
Leaving this as `systems_len()` as no strong opinions against the current name appeared.

dmlary commented:
I'm working through the merge with master to handle the `ScheduleLabel` changes for #8436 (zomg, someone else added label to schedule!!!), but I just found #7762 (interned labels) which will have the exact same merge conflicts.

For now I'm leaving this with merge conflicts.  If you're interested in merging this before #7762 I can get everything up-to-date in a couple of hours.  Just let me know if it's needed.

wilk10 commented:
AFAIK #7762 has been bumped to 0.13 today. Are there still chances for this PR to make it into 0.12?

alice-i-cecile commented:
Having looked at the milestone, this is one of a handful of ECS PRs that I'd like to push for. If #7762 is blocking this, I'm happy to prioritize that as well. It's a fine change, but by itself didn't seem critical.

dmlary commented:
I'd honestly say try to get #7762 in, as if this goes in before it, they'll have to modify stepping for the new labels.  I feel like it'll easier for me to adapt stepping for intern'd labels than for them to do it.  Although I could probably do the work under the other PR.
 
 It's up to you @alice-i-cecile. Just let me know if you want me to resolve the merge conflicts with current main.

alice-i-cecile commented:
Great, I'm going to push forward #7762 then :)

dmlary commented:
Updated for `InternedScheduleLabel` changes.

MrGVSV commented on crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:
```diff
@@ -181,6 +186,31 @@ impl SystemExecutor for MultiThreadedExecutor {
             }
         }
 
+        // If stepping is enabled, make sure we skip those systems that should
+        // not be run.
+        #[cfg(feature = "bevy_debug_stepping")]
```
Question: Why do we need to limit this block to just this feature if `skip_systems` is available as a parameter outside of it? Same for these other files.

dmlary replied:
The feature flag was requested as a mechanism to eliminate all performance impacts of stepping.  I may have gone overboard here, but as there's no other feature or functionality that is currently using the `skip_systems` argument, this felt like the best approach to meet that goal of no performance impact.

We can always remove this cfg check if another feature wants to start leveraging skip lists.

NthTensor reviewed (approved):
Looks solid. I's not entirely clear to me what parts of the example are the necessary components for using stepping, and which parts are for the example stepping interface. A more minimal example might be useful, but I guess most users won't really need to grock this.

dmlary commented on crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:
```diff
@@ -163,7 +163,12 @@ impl SystemExecutor for MultiThreadedExecutor {
+    fn run(
+        &mut self,
+        schedule: &mut SystemSchedule,
+        _skip_systems: Option<FixedBitSet>,
```
I'll make this change across all of the executors.

dmlary commented:
>  I's not entirely clear to me what parts of the example are the necessary components for using stepping, and which parts are for the example stepping interface.

You have a valid point; there is a lot of plumbing required for the breakout stepping demo just to get the system list.  I have a feeling there will quickly be a crate hiding most of the stepping details from users, but I'll take a shot at making a text-only Stepping example.

This would also be helpful for demonstrating the breakpoint functionality that isn't well demonstrated in breakout.

alice-i-cecile commented:
That sounds lovely; I think both would be valuable examples to keep.

dmlary commented:
I am working on the command-line example, but the input parsing is proving to be difficult.  Converting from `&str` into `impl IntoSystem<(), (), Marker>` has been my primary headache.  I even tried a const `Vec<(fn(), name)>`, but discovered somehow the `fn()` from that does not turn into the same System as the function itself.

I'm trying to do this without all the work that I do in the breakout stepping example to pull the `System` from the `Schedule`s to get name & `NodeId`.  I think it's going to turn into a very ugly string matching with duplicated code inside.

For reference, this is the interactive commands for the example:
```txt
Demonstration of bevy system stepping

System stepping is a tool for debugging system interactions by stepping through
their execution one at a time as the application runs.  The usage is similar to
a traditional debugger with step, continue, and breakpoint concepts.

This app reads commands line-by-line from standard input.

Commands:
    q: quit the app
    h: this help output
    empty line: run app.update() once; equivalent to rendering a single frame
    enable: enable stepping; only always_run systems will run unless step or
                continue commands are issued
    disable: disable stepping
    show: show stepping state
    b <index>, break <index>: set a breakpoint at system with index <index>
        (`Stepping::set_breakpoint`)
    a <index>, always: <index>: set a system to always run regardless if
        stepping is enabled (`Stepping::always_run()`)
    clear: clear all breakpoints
    c, continue: call `Stepping::continue_frame(); app.update()`
        This will run all systems from the stepping cursor until the end of the
        frame unless a breakpoint is encountered
    s, step: call `Stepping::step_frame(); app.update()`
        This will run the system at the stepping cursor
```

If anyone wants to help, the ideal function I've been unable to figure out would be:
```
fn parse_system_name<Marker>(name: &str) -> Option<impl IntoSystem<(), (), Marker>> {
    match name {
        "first_system" => Some(first_system),
        _ => None,
    }
}
```

The compiler has a couple of problems with it; probably the biggest is this one, which I think is telling me that `first_system` is getting wrapped into another `fn()`.  I don't know what syntax I should be using here to prevent that.
```txt
error[E0277]: the trait bound `fn() {first_system}: bevy::prelude::IntoSystem<(), (), Marker>` is not satisfied
  --> examples/ecs/system_stepping.rs:53:52
   |
53 | fn parse_system_name<Marker>(name: &str) -> Option<impl IntoSystem<(), (), Marker>> {
   |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `bevy::prelude::IntoSystem<(), (), Marker>` is not implemented for fn item `fn() {first_syst
em}`
```
If nobody has any suggestions, I'll just do an ugly match block manually calling `Stepping::set_breakpoint(Update, first_system)` for the handful of systems in the example.  Ugly, but clear.

NthTensor commented:
Oh boy, sorry I did not intend to inflict cli development on you! Given the complexity of this feature and realistically only people working on the editor are going to have to worry about it, I'd say the current example probably shouldn't block merge.

dmlary commented:
Hah, well I'm probably a good person to be doing cli programming as it's where I'm most comfortable.

I think your comment still has merit; the existing example of `Stepping` is complicated.  I think this is because it attempts to be a general purpose demonstration of stepping through any/all systems in a `Schedule`.

There's another type of demonstration where we do a more focused demonstration of simply setting a breakpoint for a system, then printing values of some resource before and after.

Perhaps this is the route I'll go down if nobody has an idea of how to implement that `parse_system_name()` function.  We already have the general purpose example in breakout.  Reimplementing it with similar complexity for the command-line doesn't really give us that much.

dmlary commented:
Ok, added a more straightforward example using the Stepping resource.  It's incredibly wordy, but exercises the resource directly in what I hope is a clear manner.

There's gonna be a few pushes to get github to run CI as https://github.com/bevyengine/bevy/issues/10902#issuecomment-1879492672 is blocking me from running CI locally.

rodolphito commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1409 @@
+            .always_run(TestSchedule, second_system)
+            .continue_frame();
+
+        // start a new frame, then run the schedule three times; first system
```
How is this running three times? I count two

dmlary replied:
Correct, updated.

rodolphito commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1409 @@
+        let mut stepping = Stepping::new();
+        stepping.add_schedule(TestSchedule).enable().step_frame();
+
+        // start a new frame, then run the schedule three times; first system
```
Three?

dmlary replied:
Lol, lesson learned.  During first pass of writing tests, never commit to a number in the comments.  I'm cleaning these all up.

rodolphito commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1409 @@
+
+    // clear all system behaviors
+    fn clear_behaviors(&mut self) {
+        self.behaviors.clear();
```
why arent `behavior_updates` cleared here too?

dmlary replied:
Great catch!  Queue Schedule changes were being applied after the clear.

I've fixed it and added a test case to cover it.  Added an additional test case to also verify the reverse order (clear_schedule(), never_run()) executes correctly.  Also confirmed the other queuing behavior (`Stepping.updates`) does not have a similar drain functionality.

rodolphito commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1409 @@
+        // cursor schedule, we'll run the schedule with the waiting action.
+        let cursor = self.cursor;
+        let (list, next_system) = if index == cursor.schedule {
+            let o = state.skipped_systems(schedule, cursor.system, self.action);
+
+            // if we just stepped this schedule, then we'll switch the action
+            // to be waiting
+            if self.action == Action::Step {
+                self.action = Action::Waiting;
+            }
+            o
```
either do this or expand `o` to something like `(skip_list, next_system)`
```suggestion
            let action = self.action;

            // if we just stepped this schedule, then we'll switch the action
            // to be waiting
            if action == Action::Step {
                self.action = Action::Waiting;
            }
            state.skipped_systems(schedule, cursor.system, action)
```

dmlary replied:
Agreed that `o` is not the appropriate variable name here.

I'm uncomfortable flipping the action change to before the call to `skipped_systems()` because it's a break from the logical order of what is happening.  I even had a moment where I missed the save to `action` and thought the change wouldn't work.

I'll split the result out into `(skip_list, next_system)` to match the `else` block, and standardize across that whole function on those names.

rodolphito reviewed (approved):
Looks to be well thought out. Love the unit tests. I'd like to see some tests with multiple copies of the same system in a schedule, and some with run conditions too.

dmlary commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1409 @@
+                _system.name()
+            );
+
+            // suppress this clippy warning; we want to keep the arms split to
```
The one I deleted once the match arms became sufficiently different.  Removing the comment.

dmlary commented:
> Looks to be well thought out. Love the unit tests. I'd like to see some tests with multiple copies of the same system in a schedule, and some with run conditions too.

This exposed a shortcoming in the `assert_schedule_runs!()` macro.  Split that macro out even further to `assert_skip_list_eq!()` for the duplicate system matching.

Added test to verify we correctly step through duplicate systems (only one instance of the system runs at a time).
Added test to verify behavior of stepping when the stepped system condition returns false (we don't run any other systems, and we step past it).

rodolphito reviewed (approved):
thanks!!

cart commented on crates/bevy_ecs/src/schedule/stepping.rs:
```diff
@@ -0,0 +1,1557 @@
+        schedule: impl ScheduleLabel,
+        system: impl IntoSystem<(), (), Marker>,
+    ) -> &mut Self {
+        let type_id = IntoSystem::into_system(system).type_id();
```
Ideally we don't need to actually construct the system to get the TypeId here. We should be able to rely on `TypeId::of::<IntoSys::System>()`. However currently that is out of sync with `FunctionSystem`'s `type_id()` implementation. I'll add a `PERF` todo.

cart replied:
Definitely not worth blocking on.

cart reviewed (approved):
I think this is in a great spot. It is minimally invasive to existing code, well tested, and behaves as expected. Great job! And sorry for the long review cycle!

cart commented on crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:
```diff
@@ -163,7 +163,12 @@ impl SystemExecutor for MultiThreadedExecutor {
+    fn run(
+        &mut self,
+        schedule: &mut SystemSchedule,
+        #[allow(unused_variables)] skip_systems: Option<FixedBitSet>,
```
Pushed!

james7132 reviewed (approved):
LGTM. It's great that  this is minimally invasive and low overhead (compared to, say, run conditions on every system), and will be super useful when used as a part of the editor.

One wishlist item from me is to support a way to step all parallel systems together, so we provide ways to try to replicate the parallel nature of the schedule. For game devs coming from an engine like Unity or Unreal, where everything is inherently serialized, may find it hard to bridge the gap to Bevy, so providing a debugging option that lets you step through groups of parallel systems may help.

cart commented:
> One wishlist item from me is to support a way to step all parallel systems together, so we provide ways to try to replicate the parallel nature of the schedule. For game devs coming from an engine like Unity or Unreal, where everything is inherently serialized, may find it hard to bridge the gap to Bevy, so providing a debugging option that lets you step through groups of parallel systems may help.

Yup this is a good idea. Fortunately I think it plays nicely into the current "skiplist" approach.
